server/utils/courseParticipationUtils.ts

```diff
@@ -3,13 +3,79 @@ import type { Request } from 'express'
 import DateUtils from './dateUtils'
 import StringUtils from './stringUtils'
 import { referPaths } from '../paths'
-import type { CourseParticipation, CourseParticipationWithName, Referral } from '@accredited-programmes/models'
+import type {
+  CourseParticipation,
+  CourseParticipationUpdate,
+  CourseParticipationWithName,
+  Referral,
+} from '@accredited-programmes/models'
 import type {
   GovukFrontendSummaryListRowWithValue,
   GovukFrontendSummaryListWithRowsWithValues,
 } from '@accredited-programmes/ui'
 
+export interface RequestWithDetailsBody extends Request {
+  body: {
+    outcome: {
+      detail: string
+      yearCompleted: string
+      yearStarted: string
+      status?: 'complete' | 'incomplete'
+    }
+    setting: {
+      communityLocation?: string
+      custodyLocation?: string
+      type?: 'community' | 'custody'
+    }
+    source: string
+  }
+}
+
 export default class CourseParticipationUtils {
+  static processedCourseDetailsFormData(request: RequestWithDetailsBody): {
+    hasFormErrors: boolean
+    submitBody: CourseParticipationUpdate
+  } {
+    const { body } = request
+    let hasFormErrors = false
+
+    const validateYear = (field: string, value: string): number | undefined => {
```

Comment:
Should probably check that this is 4 digits long and not a future year too.

Reply:
And an integer?